File: netlify/functions/news.py
import json
from datetime import datetime
from urllib.parse import parse_qs

def translate_ko(text):
    if not text: return text
    try:
        from deep_translator import GoogleTranslator
        return GoogleTranslator(source='auto', target='ko').translate(text) or text
    except Exception:
        return text

def handler(event, context):
    params = parse_qs(event.get("rawQuery") or "")
    ticker = (params.get("ticker", [""])[0] or "").upper()
    headers = {"Access-Control-Allow-Origin": "*", "Content-Type": "application/json"}
    if not ticker:
        return {"statusCode": 200, "headers": headers, "body": json.dumps({"news": []})}
    try:
        import yfinance as yf
        t = yf.Ticker(ticker)
        news = t.news or []
        result = []
        for n in news[:8]:
            c = n.get("content") or n
            title = c.get("title") or c.get("headline") or n.get("title") or ""
            link = (
                (c.get("canonicalUrl") or {}).get("url") or
                (c.get("clickThroughUrl") or {}).get("url") or
                c.get("link") or n.get("link") or "#"
            )
            publisher = (c.get("provider") or {}).get("displayName") or c.get("publisher") or ""
            pub_at = c.get("pubDate") or c.get("providerPublishTime") or n.get("providerPublishTime") or 0
            if isinstance(pub_at, str):
                try:
                    pub_at = int(datetime.fromisoformat(pub_at.replace("Z", "+00:00")).timestamp())
                except Exception:
                    pub_at = 0
            if title:
                result.append({
                    "title": translate_ko(title),
                    "titleOrig": title,
                    "link": link,
                    "publisher": publisher,
                    "publishedAt": pub_at,
                })
        return {"statusCode": 200, "headers": headers, "body": json.dumps({"news": result})}
    except Exception as e:
        return {"statusCode": 200, "headers": headers, "body": json.dumps({"news": [], "error": str(e)})}

File: netlify/functions/news.js
const https = require("https");

function get(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { timeout: 12000 }, (res) => {
      let data = "";
      res.on("data", (c) => (data += c));
      res.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch {
          reject(new Error("JSON parse error"));
        }
      });
    });
    req.on("error", reject);
    req.on("timeout", () => {
      req.destroy();
      reject(new Error("timeout"));
    });
  });
}

async function translateKo(text) {
  if (!text) return text;
  try {
    const encoded = encodeURIComponent(text);
    const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=auto&tl=ko&dt=t&q=${encoded}`;
    const data = await get(url);
    return (
      data?.[0]
        ?.map((s) => s?.[0])
        .filter(Boolean)
        .join("") || text
    );
  } catch {
    return text;
  }
}

exports.handler = async (event) => {
  const headers = {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  };
  const ticker = (event.queryStringParameters?.ticker || "").toUpperCase();
  const KEY = process.env.FINNHUB_API_KEY;

  if (!ticker || !KEY)
    return { statusCode: 200, headers, body: JSON.stringify({ news: [] }) };

  try {
    const to = Math.floor(Date.now() / 1000);
    const from = to - 30 * 24 * 3600; // 최근 30일
    const items = await get(
      `https://finnhub.io/api/v1/company-news?symbol=${ticker}&from=${new Date(from * 1000).toISOString().slice(0, 10)}&to=${new Date(to * 1000).toISOString().slice(0, 10)}&token=${KEY}`,
    );

    if (!Array.isArray(items))
      return { statusCode: 200, headers, body: JSON.stringify({ news: [] }) };

    const news = await Promise.all(
      items.slice(0, 8).map(async (n) => {
        const titleKo = await translateKo(n.headline || "");
        return {
          title: titleKo,
          titleOrig: n.headline || "",
          link: n.url || "#",
          publisher: n.source || "",
          publishedAt: n.datetime || 0,
        };
      }),
    );

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ news: news.filter((n) => n.titleOrig) }),
    };
  } catch (e) {
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ news: [], error: e.message }),
    };
  }
};
